### PR TITLE
blobstore: implement checkArchived for Ali OSS (delete-marker detection)

### PR DESCRIPTION
**Pull request by kchoy-sfdc**
## Summary

Implements `checkArchived` for the Ali OSS blob driver. When a sync `download` request has `checkArchived=true` and the underlying OSS GET fails with HTTP 404 carrying the `x-oss-delete-marker` header, the driver now lists prior versions of the key and throws `ResourceNotFoundException` populated with `ArchiveInfo` so callers can recover the archived data via the prior `versionId`.

This brings Ali to parity with AWS and GCP, which already implement the cross-cloud delete-marker / prior-version detection contract on their sync paths.

## Changes

- `AliBlobStore.handleArchivedObjects(...)` — new helper invoked from the catch blocks of all three sync `doDownload` variants (OutputStream, Path, InputStream); File and ByteArray variants delegate through them.
  - Gates on `isCheckArchived()` AND `ServiceException.statusCode == 404`.
  - Requires `x-oss-delete-marker: true` on the response (case-insensitive header lookup).
  - Calls `listObjectVersions(prefix=key, maxKeys=2)` and returns the first matching `ObjectVersion`'s id (NOT the delete marker's id, since the conformance contract re-downloads the returned id and expects original bytes).
  - Returns silently when any gate isn't met, letting the original exception propagate (never-existed key / no-bypass paths).
- `AliBlobStore.unwrapServiceException(...)` — small helper to walk the cause chain and find the underlying OSS `ServiceException` (the existing `getException` method uses a similar unwrap pattern).
- Conformance gates removed: `Assumptions.assumeFalse(ALI_PROVIDER_ID...)` dropped from `testDownload_checkArchived` and `testDownload_checkArchived_neverExisted` in `AbstractBlobStoreIT`.
- WireMock replay mappings recorded against a real versioned + WORM-enabled bucket and committed under `blob-ali/src/test/resources/mappings/`.

The async path is intentionally left unchanged for parity with AWS/GCP, which only implement this on the sync path.

## API Example

```java
BucketClient client = BucketClient.builder("ali")
    .withRegion("cn-shanghai")
    .withBucket("my-versioned-bucket")
    // ...
    .build();

try (ByteArrayOutputStream out = new ByteArrayOutputStream()) {
    client.download(
        new DownloadRequest.Builder()
            .withKey("my-key")
            .withCheckArchived(true)
            .build(),
        out);
} catch (ResourceNotFoundException e) {
    ArchiveInfo info = e.getArchiveInfo();
    if (info != null && info.isArchived()) {
        // Recover prior bytes via the version id OSS returned.
        String priorVersionId = info.getVersionId();
        // download(...) by versionId returns the original content.
    }
}
```

## Testing

- New unit tests in `AliBlobStoreTest`:
  - `testDoDownload_checkArchived_deletedOnVersionedBucket_throwsWithArchiveInfo` — positive case: delete-marker → `ArchiveInfo(archived=true, versionId=<prior>)`.
  - `testDoDownload_checkArchivedFalse_doesNotListVersionsOrChangeException` — guard is a complete no-op when the flag is off.
  - `testDoDownload_checkArchived_neverExisted_doesNotPopulateArchiveInfo` — 404 without the delete-marker header → original exception propagates, no `listObjectVersions` call.
- `mvn test -pl blob/blob-ali` — full Ali unit suite passes (191/191), checkstyle clean.
- Conformance tests `AliBlobStoreIT#testDownload_checkArchived` and `#testDownload_checkArchived_neverExisted` pass in replay mode against the recorded mappings.

### CODE REVIEW

File: blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java [hunk 1]
```diff
@@ -339,6 +350,82 @@ public DownloadResponse doDownload(DownloadRequest downloadRequest) {
...
+          && "true".equalsIgnoreCase(entry.getValue())) {
+        isDeleteMarker = true;
+        break;
+      }
+    }
+    if (!isDeleteMarker) {
+      return;
+    }
+    ListObjectVersionsResult versions = ossClient.listObjectVersions(
+        ListObjectVersionsRequest.newBuilder()
```
Comment (iamabhilaksh) [inline blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java]: Codecov is at 75% patch coverage (target 80%). The likely uncovered branches are the `versions == null` and `versions.versions() == null` guards. Two tests (`listObjectVersions` returns null result, returns empty versions) would close them and clear the gate. Non-blocking.

**Review by iamabhilaksh — APPROVED** [review]
LGTM 👍

Mirrors the AWS pattern in `AwsBlobStore.java:312-330` closely (same `maxKeys(2)`, same `versions().iterator()` approach). The exact-key match in the loop is a strict improvement over AWS — defensive against prefix collisions if they ever come up. Codecov gap inline is the only follow-up.

File: blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java [hunk 2]
```diff
@@ -339,6 +350,82 @@ public DownloadResponse doDownload(DownloadRequest downloadRequest) {
...
+            .bucket(bucket)
+            .prefix(downloadRequest.getKey())
+            .maxKeys(2L)
+            .build(),
```
Comment (SriramGuduri) [inline blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java]: question: If `listObjectVersions` throws here (network error, permission denied), does the central exception translation layer handle it gracefully? Or could it mask the original 404 + delete-marker semantics that the caller expects to receive? Asking because I was not sure if the translated exception would still convey "archived" intent to the caller. [suggestion]

File: blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java [hunk 3]
```diff
@@ -339,6 +350,82 @@ public DownloadResponse doDownload(DownloadRequest downloadRequest) {
...
+        break;
+      }
+    }
+    if (!isDeleteMarker) {
+      return;
+    }
+    ListObjectVersionsResult versions = ossClient.listObjectVersions(
+        ListObjectVersionsRequest.newBuilder()
+            .bucket(bucket)
+            .prefix(downloadRequest.getKey())
```
Comment (SriramGuduri) [inline blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java]: `maxKeys(2L)` works for the typical case (1 delete marker + 1 version = 2 entries total), but if the bucket has undergone multiple PUT/DELETE cycles, there could be multiple delete markers before the first non-marker version, causing `versions()` to be empty in this page. Consider bumping to a slightly higher value (e.g., 5 or 10) to handle repeated delete/re-upload cycles, or document the assumption that only one delete marker exists. [suggestion]

File: blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java [hunk 4]
```diff
@@ -339,6 +350,82 @@ public DownloadResponse doDownload(DownloadRequest downloadRequest) {
...
+    String versionId = null;
+    if (versions != null && versions.versions() != null) {
+      for (ObjectVersion v : versions.versions()) {
+        // Match exactly on key — the prefix listing can return sibling keys.
+        if (downloadRequest.getKey().equals(v.key())) {
+          versionId = v.versionId();
+          break;
+        }
+      }
+    }
```
Comment (SriramGuduri) [inline blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java:443]: When `versionId` ends up `null` (no matching ObjectVersion found in the listing), the thrown `ResourceNotFoundException` will have `ArchiveInfo(archived=true, versionId=null)`. The caller receiving `archived=true` but a null versionId may be confusing. Consider either not throwing (return silently and let the original 404 propagate) or documenting that a null versionId means the prior version could not be resolved. [suggestion]

**Review by SriramGuduri — APPROVED** [review]
Clean, well-documented implementation with good test coverage. The three-way gate (checkArchived flag → 404 status → delete-marker header) ensures the new logic only fires in the exact scenario it is designed for. The conformance test de-skipping confirms Ali is now at parity with AWS and GCP. Approving with minor suggestions around resilience and edge-case handling.

File: blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java [hunk 5]
```diff
@@ -339,6 +350,82 @@ public DownloadResponse doDownload(DownloadRequest downloadRequest) {
...
+    if (!isDeleteMarker) {
+      return;
+    }
+    ListObjectVersionsResult versions = ossClient.listObjectVersions(
+        ListObjectVersionsRequest.newBuilder()
+            .bucket(bucket)
+            .prefix(downloadRequest.getKey())
+            .maxKeys(2L)
+            .build(),
+        OperationOptions.defaults());
```
Comment (LihaoLiuXs) [inline blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java]: **Bug: unhandled exception from `listObjectVersions` will mask the original download failure.**

If `ossClient.listObjectVersions(...)` throws (network timeout, IAM permission denied, versioning not enabled on the bucket, OSS throttling, etc.), that exception propagates up unhandled and completely replaces the original 404 exception the caller was supposed to see. The caller gets a confusing `listObjectVersions` error instead of either:
- The `ResourceNotFoundException` with `ArchiveInfo` (happy path), or
- The original download 404 (fallback when archive detection can't complete).

This is especially problematic because the method's own Javadoc promises: *"Returns silently if [...] the original exception is allowed to propagate unchanged."* — but a `listObjectVersions` failure violates that contract.

Suggested fix — wrap in try-catch and fall through on failure:

```java
ListObjectVersionsResult versions;
try {
    versions = ossClient.listObjectVersions(
        ListObjectVersionsRequest.newBuilder()
            .bucket(bucket)
            .prefix(downloadRequest.getKey())
            .maxKeys(2L)
            .build(),
        OperationOptions.defaults());
} catch (Exception listEx) {
    // Best-effort: if version listing fails, let the original 404 propagate.
    return;
}
```

This makes the archive detection truly best-effort and prevents secondary failures from masking the primary download error.

**Review by LihaoLiuXs — COMMENTED** [review]
Overall the implementation looks solid — the guard logic, header check, and version selection are correct. One critical concern about unhandled exceptions from the `listObjectVersions` call that could mask the original download failure at runtime.

File: blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java (hunk 5 above)
Comment (kchoy-sfdc) [reply blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java]: Good catch — you're right that this violated the method's own best-effort contract. Fixed in `1bb18dc8`: the version lookup is now wrapped so that any failure (network error, permission denied, versioning disabled, throttling) falls through and lets the caller's original download exception propagate unchanged, rather than masking it with a secondary error. While addressing this I also made the version resolution deterministic — see the reply on the `maxKeys` thread. Added a unit test (`testDoDownload_checkArchived_listVersionsFails_propagatesOriginalException`) asserting the original exception surfaces and no `ResourceNotFoundException` is produced.

File: blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java (hunk 2 above)
Comment (kchoy-sfdc) [reply blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java]: Confirmed your concern was valid — a `listObjectVersions` failure would have replaced the original 404 + delete-marker semantics the caller expects. Fixed in `1bb18dc8` by making archive detection strictly best-effort: on any listing failure the guard returns and the original exception propagates unchanged, so the "archived" intent is preserved on the happy path and never clobbered by a secondary error on the failure path.

File: blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java (hunk 3 above)
Comment (kchoy-sfdc) [reply blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java]: Great point, and it pushed me to fix this properly rather than just bump the constant. The root issue is that OSS delete markers share the page-size budget with real versions, so *any* fixed `maxKeys` can be defeated by enough stacked markers from repeated PUT/DELETE cycles. Rather than guess a larger number, `1bb18dc8` switches to the SDK paginator (`listObjectVersionsPaginator`) and walks pages until the first exact-key match — deterministic regardless of how many delete markers precede the version, with no magic number. Added `testDoDownload_checkArchived_versionOnLaterPage_resolvesAcrossPages` to lock in cross-page resolution.

File: blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java (hunk 4 above)
Comment (kchoy-sfdc) [reply blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java:443]: Agreed — reporting `archived=true` with a null `versionId` the caller can't act on is confusing. Fixed in `1bb18dc8`: if no prior version can be resolved, the guard now falls through and lets the original 404 propagate instead of throwing with a null `versionId`. Combined with the pagination change this should be rare, but it's handled explicitly. Covered by `testDoDownload_checkArchived_noMatchingVersion_propagatesOriginalException` and `testDoDownload_checkArchived_emptyVersionsPage_propagatesOriginalException`.

File: blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java (hunk 1 above)
Comment (kchoy-sfdc) [reply blob/blob-ali/src/main/java/com/salesforce/multicloudj/blob/ali/AliBlobStore.java]: Thanks — the new unit tests for the listing-failure, empty-versions-page, and no-matching-version branches exercise exactly the guards that were uncovered, which should close the patch-coverage gap. The recordings were regenerated for the paginated lookup (`26981f28`) and the conformance tests pass in replay.

**Review by LihaoLiuXs — APPROVED** [review]
✅ Approve. All prior review comments properly addressed:

1. ✅ **Unhandled exception from `listObjectVersions`** — Now caught by `catch (Exception listFailure)` with best-effort semantics: original download failure preserved, listing failure logged but not propagated.

2. ✅ **`maxKeys(2L)` pagination concern** — Replaced with SDK paginator (`listObjectVersionsPaginator`) that pages through results transparently. Comment explains why bounded pages were insufficient (delete markers share page budget).

3. ✅ **Null `versionId` in exception** — Now falls through if no matching version found (doesn't throw with null data).

Architecture is correct: `handleArchivedObjects` is best-effort, correctly placed after the 404-status check, and uses paginated listing to find the prior version. 7 new test cases cover all edge cases including listing failure, empty versions page, and no-matching-version branches.
